experimental/iterators/lib/Dialects/Tabular/IR/Tabular.cpp

```diff
@@ -0,0 +1,124 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "iterators/Dialect/Tabular/IR/Tabular.h"
+
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::iterators;
+
+//===----------------------------------------------------------------------===//
+// Tabular dialect
+//===----------------------------------------------------------------------===//
+
+#include "iterators/Dialect/Tabular/IR/TabularOpsDialect.cpp.inc"
+
+void TabularDialect::initialize() {
+#define GET_OP_LIST
+  addOperations<
+#include "iterators/Dialect/Tabular/IR/TabularOps.cpp.inc"
+      >();
+  addTypes<
+#define GET_TYPEDEF_LIST
+#include "iterators/Dialect/Tabular/IR/TabularOpsTypes.cpp.inc"
+      >();
+}
+
+//===----------------------------------------------------------------------===//
+// Tabular interfaces
+//===----------------------------------------------------------------------===//
+
+#include "iterators/Dialect/Tabular/IR/TabularOpInterfaces.cpp.inc"
+#include "iterators/Dialect/Tabular/IR/TabularTypeInterfaces.cpp.inc"
+
+//===----------------------------------------------------------------------===//
+// Tabular operations
+//===----------------------------------------------------------------------===//
+
+#define GET_OP_CLASSES
+#include "iterators/Dialect/Tabular/IR/TabularOps.cpp.inc"
+
+LogicalResult CreateTabularViewOp::verify() {
+  auto viewType = view().getType().cast<TabularViewType>();
+  TypeRange columnTypes = viewType.getColumnTypes();
+
+  // Verify matching column/element types.
+  if (!llvm::all_of_zip(memrefs().getTypes(), columnTypes,
+                        [](Type t1, Type t2) {
+                          return t1.cast<MemRefType>().getElementType() == t2;
+                        })) {
+
+    std::string viewColumnTypes;
+    {
+      llvm::raw_string_ostream stream(viewColumnTypes);
+      llvm::interleaveComma(columnTypes, stream,
+                            [&](Type type) { type.print(stream); });
+    }
+    std::string memrefColumnTypes;
+    {
+      llvm::raw_string_ostream stream(memrefColumnTypes);
+      llvm::interleaveComma(memrefs().getTypes(), stream, [&](Type type) {
+        type.cast<MemRefType>().getElementType().print(stream);
+      });
+    }
+    return emitOpError()
+           << "type mismatch: should return a tabular view with the column "
+           << "types of the elements of the input memrefs (namely: "
+           << "tabular_view<" << memrefColumnTypes << ">) "
+           << "but returns a tabular_view<" << viewColumnTypes << ">.";
+  }
+
+  // Verify memrefs have rank 1 and contiguous memory layout.
+  {
+    for (const auto &indexedMemref : llvm::enumerate(memrefs().getTypes())) {
```

Comment:
You can use StridedMemRefRankOf mlir/include/mlir/IR/OpBase.td and trim most of this out.
Feel free to add extra support for contiguous and what else may be missing for your use case and send it for review upstream.

Reply:
OK, I added some tablegen constructs to `TabularOps.td` in the commit "Move some type constraints to tablegen code." What do you think? Is this worth/ready for upstreaming?